### ISSUE TITLE
Question about the alignment loss (eq.2) and uniformity loss (eq.3)

### ISSUE THREAD

**Issue by RenShuhuai-Andy** [princeton-nlp/SimCSE#41]
Hi, thank you for your excellent work.
I try to calculate the alignment loss (eq.2) and uniformity loss (eq.3) on the STS-B dataset, and re-implement the results in figure 2.
Referring to the instructions in your paper and the code in [https://github.com/SsnL/align_uniform/](https://github.com/SsnL/align_uniform/), I import the following functions to calculate the two losses:
```
def align_loss(x, y, alpha=2):
    return (x - y).norm(p=2, dim=1).pow(alpha).mean()

def uniform_loss(x, t=2):
    return torch.pdist(x, p=2).pow(2).mul(-t).exp().mean().log()
```
and modify the code from [https://github.com/princeton-nlp/SimCSE/blob/main/SentEval/senteval/sts.py#L59-L85](https://github.com/princeton-nlp/SimCSE/blob/main/SentEval/senteval/sts.py#L59-L85) to 
```
def run(self, params, batcher):
    results = {}
    all_sys_scores = []
    all_gs_scores = []
    ################# newly added
    all_loss_align = []
    all_loss_uniform = []
    #################
    for dataset in self.datasets:
        sys_scores = []
        input1, input2, gs_scores = self.data[dataset]
        for ii in range(0, len(gs_scores), params.batch_size):
            batch1 = input1[ii:ii + params.batch_size]
            batch2 = input2[ii:ii + params.batch_size]
            batch_gs_scores = gs_scores[ii:ii + params.batch_size]  # newly added

            # we assume get_batch already throws out the faulty ones
            if len(batch1) == len(batch2) and len(batch1) > 0:
                enc1 = batcher(params, batch1)
                enc2 = batcher(params, batch2)
                ################# newly added
                pos_indices = [i for i in range(len(batch_gs_scores)) if batch_gs_scores[i] >= 4.0]
                enc1_pos = enc1[pos_indices]
                enc2_pos = enc2[pos_indices]
                loss_align = align_loss(enc1_pos, enc2_pos)
                loss_uniform = uniform_loss(torch.cat((enc1, enc2), dim=0))
                all_loss_align.append(loss_align)
                all_loss_uniform.append(loss_uniform)
                ################# 

                for kk in range(enc2.shape[0]):
                    sys_score = self.similarity(enc1[kk], enc2[kk])
                    sys_scores.append(sys_score)
        all_sys_scores.extend(sys_scores)
        all_gs_scores.extend(gs_scores)
        results[dataset] = {'pearson': pearsonr(sys_scores, gs_scores),
                            'spearman': spearmanr(sys_scores, gs_scores),
                            'nsamples': len(sys_scores),
                            'align_loss': np.mean(all_loss_align),  # newly added
                            'uniform_loss': np.mean(all_loss_uniform)}  # newly added
        logging.debug('%s : pearson = %.4f, spearman = %.4f, align_loss = %.4f, uniform_loss = %.4f' %
                      (dataset, results[dataset]['pearson'][0],
                       results[dataset]['spearman'][0], results[dataset]['align_loss'],
                       results[dataset]['uniform_loss']))
```
However, my results (align_loss and uniform_loss) are:
![image](https://user-images.githubusercontent.com/30052148/121996125-a8978900-cdda-11eb-84f7-6f5adaa4195f.png)
which is inconsistent with the results in figure 2 (about 0.35 for align_loss and about -3.4 for uniform_loss).
I wonder if I make any mistakes? Thanks.

**Comment by yaoxingcheng** [comment]
Hi, thanks for your attention to our work. 
In our paper, all embeddings are l2-normalized before computing the alignment and uniformity metrics. I think you should normalize the embeddings to unit vectors first before calculating the losses.

**Comment by RenShuhuai-Andy** [comment]
Got it, thanks.
